e3nn_jax/_src/spherical_signal.py

```diff
@@ -0,0 +1,215 @@
+        """Rotate the signal by the given quaternion."""
+        coeffs = self._coeffs(lmax)
+        transformed_coeffs = coeffs.transform_by_quaternion(q, k)
+        transformed_grid_values = e3nn.to_s2grid(transformed_coeffs, *self.grid_resolution(), quadrature=self.quadrature)
+        return SphericalSignal(
+            transformed_grid_values, self.y, self.alpha, self.quadrature, self.quadrature_weights, self.p_val, self.p_arg
+        )
+
+    def apply(self, func: Callable[[chex.Array], chex.Array]):
+        """Applies a function pointwise on the grid."""
+        return SphericalSignal(func(self.grid_values), self.y, self.alpha, self.quadrature, self.quadrature_weights)
+
+    @staticmethod
+    def irreps_for_spherical_signal(lmax: int, p_val: int, p_arg: int) -> e3nn.Irreps:
+        """Returns all Irreps upto l = lmax and of the required parity."""
+        return e3nn.Irreps([(1, (l, p_val * p_arg**l)) for l in range(lmax + 1)])
+
+    @classmethod
+    def from_irreps_array(
+        cls, coeffs: e3nn.IrrepsArray, res_beta: int = 100, res_alpha: int = 99, quadrature: str = "gausslegendre"
+    ) -> "SphericalSignal":
+        """Creates a SphericalSignal from an IrrepsArray and a given grid resolution."""
+        grid_values = e3nn.to_s2grid(coeffs, res_beta=res_beta, res_alpha=res_alpha, quadrature=quadrature)
+        s2grid = e3nn.s2grid(res_alpha=res_alpha, res_beta=res_beta, quadrature=quadrature)
+        y, alpha, quadrature_weights = jax.tree_map(jnp.asarray, s2grid)
+        return cls(grid_values, y, alpha, quadrature, quadrature_weights)
+
+    @staticmethod
+    def _find_peaks_2d(x: np.ndarray) -> List[Tuple[int, int]]:
+        """Helper for finding peaks in a 2D signal."""
+        iii = []
+        for i in range(x.shape[0]):
+            jj, _ = scipy.signal.find_peaks(x[i, :])
+            iii += [(i, j) for j in jj]
+
+        jjj = []
+        for j in range(x.shape[1]):
+            ii, _ = scipy.signal.find_peaks(x[:, j])
+            jjj += [(i, j) for i in ii]
+
+        return list(set(iii).intersection(set(jjj)))
+
+    def find_peaks(self, lmax: int):
+        r"""Locate peaks on the signal on the sphere."""
+        grid_resolution = self.grid_resolution()
+        x1, f1 = e3nn.s2grid_vectors(self.y, self.alpha), self.grid_values
+
+        # Rotate signal.
+        abc = np.array([jnp.pi / 2, jnp.pi / 2, jnp.pi / 2])
+        R = e3nn.angles_to_matrix(*abc)
+        r_signal = self.transform_by_angles(*abc, lmax=lmax)
+        x2, f2 = np.einsum("ij,baj->bai", R.T, x1), r_signal.grid_values
+
+        ij = self._find_peaks_2d(f1)
+        x1p = np.stack([x1[i, j] for i, j in ij])
+        f1p = np.stack([f1[i, j] for i, j in ij])
+
+        ij = self._find_peaks_2d(f2)
+        x2p = np.stack([x2[i, j] for i, j in ij])
+        f2p = np.stack([f2[i, j] for i, j in ij])
+
+        # Union of the results
+        mask = scipy.spatial.distance.cdist(x1p, x2p) < 2 * jnp.pi / max(*grid_resolution)
```

Comment:
Are you really mixing `scipy` and `jax` ? No good :)

Reply:
Sorry, it is just jnp.pi. Will fix.

Reply:
Oh true, I don't know why I thought that the rest of the function was also in jax